num/__private/signed_integer_macros.h

```diff
@@ -0,0 +1,445 @@
+constexpr inline bool can_sub_without_overflow(const T& l, const T& r) {
+  if (r.primitive_value <= 0)
+    return l.primitive_value <= T::MAX_PRIMITIVE + r.primitive_value;
+  else
+    return l.primitive_value >= T::MIN_PRIMITIVE + r.primitive_value;
+}
+
+template <class T>
+constexpr inline bool can_mul_without_overflow(const T& l, const T& r) {
+  // TODO: Is this really correct always? False positives?
+  if (l.primitive_value == 0 || r.primitive_value == 0) {
+    return true;
+  } else if (l.primitive_value > 0 && r.primitive_value > 0) {
+    return l.primitive_value <= T::MAX_PRIMITIVE / r.primitive_value;
+  } else if (r.primitive_value > 0) {
+    return l.primitive_value >= T::MIN_PRIMITIVE / r.primitive_value;
+  } else {
+    return r.primitive_value >= T::MIN_PRIMITIVE / l.primitive_value;
+  }
+}
+
+template <class T>
+constexpr inline bool can_div_without_overflow(const T& l, const T& r) {
+  if (r.primitive_value == 0) {
+    // Divide by zero.
+    return false;
+  } else if (r.primitive_value == -1) {
+    // Overflow: MIN / -1 => MAX + 1.
+    return l.primitive_value > T::MIN();
+  } else {
+    return true;
+  }
+}
+
+}  // namespace sus::num::__private
+
+#define _sus__signed_constants(T, Max, Bits)                         \
+  static constexpr auto MIN_PRIMITIVE = primitive_type{-Max - 1};    \
+  static constexpr auto MAX_PRIMITIVE = primitive_type{Max};         \
+  static constexpr inline T MIN() noexcept { return MIN_PRIMITIVE; } \
+  static constexpr inline T MAX() noexcept { return MAX_PRIMITIVE; } \
+  static constexpr inline /* TODO: u32 */ uint32_t BITS() noexcept { \
+    return /* TODO: u32(*/ Bits /*)*/;                               \
+  }                                                                  \
+  static_assert(true)
+
+#define _sus__signed_impl(T, LargerT) \
+  _sus__signed_integer_comparison(T); \
+  _sus__signed_unary_ops(T);          \
+  _sus__signed_binary_logic_ops(T);   \
+  _sus__signed_binary_bit_ops(T);     \
+  _sus__signed_mutable_logic_ops(T);  \
+  _sus__signed_mutable_bit_ops(T);    \
+  _sus__signed_abs(T);                \
+  _sus__signed_add(T);                \
+  _sus__signed_div(T);                \
+  _sus__signed_mul(T, LargerT)
+
+#define _sus__signed_integer_comparison(T)                                    \
+  /** sus::concepts::Eq<##T##> trait. */                                      \
+  friend constexpr inline bool operator==(const T& l, const T& r) noexcept {  \
+    return (l.primitive_value <=> r.primitive_value) == 0;                    \
+  }                                                                           \
+  /** sus::concepts::Ord<##T##> trait. */                                     \
+  friend constexpr inline auto operator<=>(const T& l, const T& r) noexcept { \
+    return l.primitive_value <=> r.primitive_value;                           \
+  }                                                                           \
+  static_assert(true)
```

Comment:
Why are these asserts necessary?

Reply:
They let you write a `;` at the end of the instantiation of the macro - which lets code formatting work better.

Reply:
It's like do { .. } while(0) in macros but for inside a class decl instead.